### ISSUE TITLE
Interacting with chdb via clickhouse-go

### ISSUE THREAD

**Issue by iyuroch** [chdb-io/chdb-go#6]
Is there a way to connect to chdb using clickhouse-go?

**Comment by auxten** [comment]
As chDB is now part of ClickHouse. I think this is the next step we will do.

**Comment by arnaudbriche** [comment]
Any update on this ? Is it planned or not yet decided ?

**Comment by auxten** [comment]
Yes, it's planed. Will do it on Q2 2025

**Comment by orloffv** [comment]
@auxten any news here?

**Comment by auxten** [comment]
We are a little behind schedule.
Can you tell us more about your use case or specific feature you guys want?

**Comment by orloffv** [comment]
We have an application that uses the clickhouse-go library to work with Clickhouse. I would like to use chdb for unit tests.

**Comment by daviddavid533** [comment]
news?

**Comment by aybabtme** [comment]
@auxten similar here to @orloffv it would be great to be able to use `chdb` in unit tests for debugging, as the lightweight and file approach makes it easy to leave the DB file behind to debug failed tests.

This is much more cumbersome to achieve using the full clickhouse server.

Also the code we use for interacting with clickhouse is modular but has a dependency on the `clickhouse-go`'s insertion mechanisms.

**Comment by auxten** [comment]
@s0und0fs1lence how is your idea?
Anyone interested to make a contribution?

**Comment by s0und0fs1lence** [comment]
We can try to bind a unix socket to the underlying clickhouse instance and wrap it into chdb-go.
This has 2 approach (both feasible  but with 2 different required time):


1. Rewrite the clickhouse tcp handler in plain go directly inside chdb-go
2. Modify the cpp chdb library to expose the tcp socket trough new functions

The first approach will be the most "indipendent", because it will not require to change the cpp codebase, but as disadvantage will require every other binding to make the same. The later will be more portable and in my opinion faster to develop, given the fact that we don't have to handle the binary decoding/encoding and we can just relay on the clickhouse one.

@auxten do you have room on your team for making the necessary changes in the chdb library? 
Otherwise, i can try to develop a poc in pure go and try to work around the problem.

Let me know

**Comment by auxten** [comment]
@s0und0fs1lence  I think your proposal is feasible. In addition, the connection implemented by chDB is similar to what you said to start a clickhouse server, but it does not start a new process, and does not go to the network, thus avoiding a copy. You can realize the plan you said based on the connection of chDB.
It seems better to make a PR to clickhouse-go to support chDB as an alternative

**Comment by auxten** [comment]
@s0und0fs1lence BTW, the chDB v3.6.0 released with static libchdb support. Maybe you can have a try. https://github.com/chdb-io/chdb/pull/362

**Comment by s0und0fs1lence** [comment]
Yes, the real problem is that _clickhouse-go_ sends the binary protocol to the fd, so we need to decode it before sending the query trough the connection. It also expect the binary protocol in response, so we have to encode it accordingly. 

Can we expose at least 2 new functions that accepts the binary message and does all the decoding and then return the response?

**Comment by s0und0fs1lence** [comment]
Nevermind, 
I just remembered that there is the _ch-go_ library which implements the clickhouse binary format in go.
I'll start the implementation around this in the next days.

**Comment by orloffv** [comment]
@s0und0fs1lence do you have any news?
